### PR TITLE
feat(leveling): implement leveling path code encode/decode

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- Implement shareable leveling path codes with protobuf serialization
- Add version header for forward compatibility ("LP" magic + uint16 version)
- Include comprehensive validation and error handling
- Create unit tests covering all edge cases

## Changes

- `src/utils/levelingPathCodec.ts` - Core encode/decode implementation
- `test/unit/utils/levelingPathCodec.test.ts` - 33 unit tests

## Technical Details

**Code Format:**
1. Serialize LevelingPath to protobuf binary
2. Prepend header (magic "LP" + version)
3. Compress with zlib deflate
4. Encode to URL-safe base64

**Functions:**
- `encodeLevelingPath(path, options)` - Encode path to shareable code
- `decodeLevelingPath(code)` - Decode code back to LevelingPath
- `getCodeVersion(code)` - Check version without full decode
- `canDecode(code)` - Verify compatibility

## Test plan

- [x] Roundtrip encoding/decoding preserves all data
- [x] All character classes preserved correctly
- [x] All step trigger/action types preserved
- [x] Unicode characters in name/notes preserved
- [x] Invalid format returns appropriate error
- [x] Invalid base64 returns decode error
- [x] Non-zlib data returns decompress error
- [x] Invalid header returns header error
- [x] Unsupported version returns version error
- [x] Missing required field returns validation error
- [x] Performance: 100 encodes/decodes < 1000ms
- [x] Lint passes
- [x] Build succeeds

Closes https://github.com/polaz/pob2-web/issues/69

### CODE REVIEW

File: src/utils/levelingPathCodec.ts [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+ * Uses browser's btoa with proper binary handling.
+ */
+function uint8ArrayToBase64(bytes: Uint8Array): string {
+  let binary = '';
+  const len = bytes.length;
+  for (let i = 0; i < len; i++) {
+    binary += String.fromCharCode(bytes[i] as number);
+  }
+  return btoa(binary);
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Added the performance explanation comment matching the existing pattern in buildCode.ts. The for-loop approach avoids the overhead of Array.from() and join() which would create intermediate arrays.

File: src/utils/levelingPathCodec.ts [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+  header[2] = version & 0xff;
+  header[3] = (version >> 8) & 0xff;
+  return header;
+}
+
+/**
+ * Parse and validate header bytes.
+ * Returns version number if valid, null if invalid magic.
+ */
+function parseHeader(data: Uint8Array): { valid: boolean; version: number } {
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Explicit return type is already present in the function signature. The function declaration includes the return type annotation.

File: src/utils/levelingPathCodec.ts [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+    binary += String.fromCharCode(bytes[i] as number);
+  }
+  return btoa(binary);
+}
+
+/**
+ * Decode base64 string to Uint8Array.
+ * Uses browser's atob with proper binary handling.
+ */
+function base64ToUint8Array(base64: string): Uint8Array {
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Explicit return type is already present in the function signature. The function declaration includes the return type annotation.

File: src/utils/levelingPathCodec.ts [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+    bytes[i] = binary.charCodeAt(i);
+  }
+  return bytes;
+}
+
+/**
+ * Convert standard base64 to URL-safe base64.
+ * Replaces + with - and / with _
+ */
+function toUrlSafeBase64(base64: string): string {
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Explicit return type is already present in the function signature. The function declaration includes the return type annotation.

File: src/utils/levelingPathCodec.ts [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+    return false;
+  }
+  return CODE_PATTERN.test(code);
+}
+
+/**
+ * Validate that a decoded path has required fields.
+ * LevelingPath requires an id field.
+ */
+function validatePath(path: LevelingPathType): LevelingPathCodeError | null {
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Explicit return type is already present in the function signature. The function declaration includes the return type annotation.

File: test/unit/utils/levelingPathCodec.test.ts [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+
+function createMinimalPath(): LevelingPath {
+  return {
+    id: 'test-path-1',
+    checkpoints: [],
+    steps: [],
+  };
+}
+
+function createFullPath(): LevelingPath {
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/utils/levelingPathCodec.test.ts]: Explicit return type is already present in the function signature. The function declaration includes the return type annotation.

File: src/utils/levelingPathCodec.ts [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+
+// =============================================================================
+// Header Utilities
+// =============================================================================
+
+/**
+ * Create header bytes with magic and version.
+ * Format: [0x4C, 0x50, version_low, version_high]
+ */
+function createHeader(version: number): Uint8Array {
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Explicit return type is already present in the function signature. The function declaration includes the return type annotation.

File: src/utils/levelingPathCodec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Added @param and @returns JSDoc documentation to clarify the function signature. The explicit return type was already present in the TypeScript declaration.

File: src/utils/levelingPathCodec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Added @param and @returns JSDoc documentation to clarify the function signature. The explicit return type was already present in the TypeScript declaration.

File: test/unit/utils/levelingPathCodec.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/utils/levelingPathCodec.test.ts]: Added @param and @returns JSDoc documentation to clarify the function signature. The explicit return type was already present in the TypeScript declaration.

File: src/utils/levelingPathCodec.ts [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+
+// =============================================================================
+// Base64 Encoding/Decoding
+// =============================================================================
+
+/**
+ * Encode Uint8Array to base64 string.
+ * Uses browser's btoa with proper binary handling.
+ */
+function uint8ArrayToBase64(bytes: Uint8Array): string {
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: Added @param and @returns JSDoc documentation to clarify the function signature. The explicit return type was already present in the TypeScript declaration.

File: test/unit/utils/levelingPathCodec.test.ts [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+
+      expect(result.success).toBe(false);
+      if (!result.success) {
+        expect(result.error.type).toBe('INVALID_FORMAT');
+      }
+    });
+
+    it('should return error for invalid base64', () => {
+      // Valid format but garbage content (!! are invalid base64)
+      const result = decodeLevelingPath('AAAAAAAAAA!!!');
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/utils/levelingPathCodec.test.ts]: Fixed. Changed test to use 'eNrtWW1v2zgMA' which passes URL-safe format validation but fails at decode/decompress stage. Updated comment to reflect the actual test behavior.

File: test/unit/utils/levelingPathCodec.test.ts [hunk 10]
```diff
@@ -0,0 +1,599 @@
+/**
+ * Unit tests for leveling path code encode/decode functionality.
+ *
+ * Tests cover:
+ * - Roundtrip encoding/decoding
+ * - Header validation (magic bytes, version)
+ * - Error handling for invalid codes
+ * - Performance requirements (<100ms)
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/utils/levelingPathCodec.test.ts]: Fixed. Updated header comment to '<1000ms for 100 operations>' to match the actual test implementation.

File: src/utils/levelingPathCodec.ts [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+  } catch (e) {
+    const error: LevelingPathCodeError = {
+      type: 'DECODE_FAILED',
+      message: `Failed to decode base64: ${e instanceof Error ? e.message : 'Unknown error'}`,
+    };
+    // Conditional spread to avoid assigning undefined with exactOptionalPropertyTypes
+    if (e instanceof Error) {
+      error.cause = e;
+    }
+    return { success: false, error };
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/utils/levelingPathCodec.ts]: The ternary approach `cause: e instanceof Error ? e : undefined` would fail compilation with `exactOptionalPropertyTypes` enabled - TypeScript does not allow assigning `undefined` to an optional property. The conditional assignment pattern is the correct approach for this strict mode setting. Added explicit comment explaining this constraint.

File: test/unit/utils/levelingPathCodec.test.ts [hunk 12]
```diff
@@ -0,0 +1,638 @@
...
+      // Create a valid compressed payload but with wrong magic bytes
+      // Add padding to ensure output is long enough (>10 chars after base64)
+      const wrongHeader = new Uint8Array([
+        0x00, 0x00, 0x01, 0x00, // Wrong magic (not "LP")
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Padding
+      ]);
+      const compressed = deflateSync(wrongHeader);
+      const code = btoa(String.fromCharCode(...compressed))
+        .replace(/\+/g, '-')
+        .replace(/\//g, '_');
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/utils/levelingPathCodec.test.ts]: Refactored to use `encodeUrlSafeBase64` from the shared `src/utils/base64.ts` module instead of manual base64 encoding. Tests now use the actual utility function.
